sdk/go/common/resource/testing/rapid.go

```diff
@@ -0,0 +1,226 @@
+//nolint:lll
+package testing
+
+import (
+	"github.com/stretchr/testify/require"
+	"pgregory.net/rapid"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+// TypeGenerator generates legal tokens.Type values.
+func TypeGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) tokens.Type {
+		return tokens.Type(rapid.StringMatching(`^[a-zA-Z][-a-zA-Z0-9_]*:([^0-9][a-zA-Z0-9._/]*)?:[^0-9][a-zA-Z0-9._/]*$`).Draw(t, "type token").(string))
+	})
+}
+
+// URNGenerator generates legal resource.URN values.
+func URNGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.URN {
+		stackName := tokens.QName(rapid.StringMatching(`^((:[^:])[^:]*)*:?$`).Draw(t, "stack name").(string))
+		projectName := tokens.PackageName(rapid.StringMatching(`^((:[^:])[^:]*)*:?$`).Draw(t, "project name").(string))
+		parentType := TypeGenerator().Draw(t, "parent type").(tokens.Type)
+		resourceType := TypeGenerator().Draw(t, "resource type").(tokens.Type)
+		resourceName := tokens.QName(rapid.StringMatching(`^((:[^:])[^:]*)*:?$`).Draw(t, "resource name").(string))
+		return resource.NewURN(stackName, projectName, parentType, resourceType, resourceName)
+	})
+}
+
+// IDGenerator generates legal resource.ID values.
+func IDGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.ID {
+		return resource.ID(rapid.String().Draw(t, "ids").(string))
+	})
+}
+
+// SemverStringGenerator generates legal semver strings.
+func SemverStringGenerator() *rapid.Generator {
+	return rapid.StringMatching(`^v?(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`)
+}
+
+// UnknownPropertyGenerator generates the unknown resource.PropertyValue.
+func UnknownPropertyGenerator() *rapid.Generator {
+	return rapid.Custom(func(t *rapid.T) resource.PropertyValue {
```

Comment:
Could use `rapid.Just(resource.MakeComputed(resource.NewStringProperty("")))`? The outer Custom wrapper it seems to me makes it go over fresh new values instead of reusing a single value. The difference wouldn't be observable if consumers never used pointer equality on the value, but maybe they do.

Reply:
I've used a `Custom` wrapper here to ensure that all of the `PropertyGenerator` functions return a `Generator` that returns values of type `resource.PropertyValue`. This would be clearer if we had generics (indeed, I sort of want to fork `rapid` and port it to use Go generics just to see what the usability is like).